### PR TITLE
fix(access-token): lock the jsonlines database

### PR DESCRIPTION
**Pull request by michaelfig**
## Description

If there were multiple simultaneous calls to `getAccessToken`, they would each commit the token database with a different fresh token, and the last commit would win.

That would mean at least one of the callers would return an invalid, stale token to its caller.

This PR introduces a file lock to prevent the conflict from occurring.

### CODE REVIEW

File: packages/access-token/test/test-token.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+const tmpDir = prefix =>
+  new Promise((resolve, reject) => {
+    tmp.dir({ unsafeCleanup: true, prefix }, (err, name, removeCallback) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve([name, removeCallback]);
+      }
+    });
+  });
```
Comment (mhofman) [inline packages/access-token/test/test-token.js]: We really should abstract this somewhere. We have about 4 different versions of about the same code

File: packages/access-token/src/access-token.js [hunk 2]
```diff
@@ -45,15 +54,27 @@ export async function getAccessToken(
...
-  const { storage, commit, close } = openJSONStore(sharedStateDir);
-  const accessTokenKey = `accessToken/${port}`;
-  if (!storage.has(accessTokenKey)) {
-    storage.set(accessTokenKey, await generateAccessToken());
-    await commit();
+  let accessToken;
+
+  const lockFilePath = path.join(sharedStateDir, DATA_FILE);
+  const release = await lockfile.lock(sharedStateDir, {
+    lockFilePath,
```
Comment (mhofman) [inline packages/access-token/src/access-token.js]: This feels a little leaky. Can we instead add a function returned by `openJSONStore` to `acquireLock` / "start a transaction" ? Then the lock can be automatically released if taken when `close`-ing the store.

File: packages/access-token/src/json-store.js [hunk 3]
```diff
@@ -9,7 +9,7 @@ import Readlines from 'n-readlines';
 //
 // For compatibility, because it's tricky to diagnose that the file location has
 // changed when you're just trying to use the wallet from the browser.
-const DATA_FILE = 'swingset-kernel-state.jsonlines';
+export const DATA_FILE = 'swingset-kernel-state.jsonlines';
```
Comment (mhofman) [inline packages/access-token/src/json-store.js]: I'd prefer not exporting this implementation details and instead have the locking mechanism implemented in the store itself.

File: packages/access-token/tsconfig.json [hunk 4]
```diff
@@ -1,6 +1,9 @@
 // This file can contain .js-specific Typescript compiler config.
 {
   "extends": "../../tsconfig.json",
+  "compilerOptions": {
+    "allowSyntheticDefaultImports": true
```
Comment (mhofman) [inline packages/access-token/tsconfig.json:5]: Why is this needed ?

File: yarn.lock [hunk 5]
```diff
@@ -5211,6 +5211,11 @@ graceful-fs@^4.1.11, graceful-fs@^4.1.15, graceful-fs@^4.1.2, graceful-fs@^4.1.6
   resolved "https://registry.yarnpkg.com/graceful-fs/-/graceful-fs-4.2.10.tgz#147d3a006da4ca3ce14728c7aefc287c367d7a6c"
   integrity sha512-9ByhssR2fPVsNZj478qUUbKfmL0+t5BDVyjShtyZZLiK7ZDAArFFfopyOTj0M05wE2tJPisA4iTnnXl2YoPvOA==
 
+graceful-fs@^4.2.4:
```
Comment (mhofman) [inline yarn.lock]: please dedup

**Review by mhofman — COMMENTED** [review]
Please let me know what you think about moving the locking responsibilities.

File: packages/access-token/src/access-token.js (hunk 2 above)
Comment (michaelfig) [reply packages/access-token/src/access-token.js]: The problem is that `openJSONStore` itself must take the lock, or else the `.has` and `.get` can be out of sync with what's on disk.  I could add `openJSONStore(dir, exclusive = false)` where `exclusive = true` would do the semantics I need and not be a leaky abstraction.

File: packages/access-token/tsconfig.json (hunk 4 above)
Comment (michaelfig) [reply packages/access-token/tsconfig.json:5]: It's needed to silence my IDE from complaining about `import path from 'path'` and everything like it.

File: packages/access-token/src/access-token.js (hunk 2 above)
Comment (michaelfig) [reply packages/access-token/src/access-token.js:61]: Okay, I see away to do `acquireLock` now.

File: packages/access-token/package.json [hunk 6]
```diff
@@ -14,11 +14,14 @@
...
     "lint": "run-s --continue-on-error lint:*",
     "lint-fix": "yarn lint:eslint --fix",
-    "lint:eslint": "eslint ."
+    "lint:eslint": "eslint .",
+    "lint:types": "tsc"
   },
   "dependencies": {
-    "@agoric/assert": "^0.6.0",
+    "@types/n-readlines": "^1.0.3",
+    "@types/proper-lockfile": "^4.1.2",
```
Comment (mhofman) [inline packages/access-token/package.json]: Shouldn't these be `devDependencies` (sorry missed it the first time around) ?

File: packages/access-token/src/access-token.js [hunk 7]
```diff
@@ -5,7 +5,12 @@ import path from 'path';
 
 import { openJSONStore } from './json-store.js';
 
-// Adapted from https://stackoverflow.com/a/43866992/14073862
+/**
+ * Adapted from https://stackoverflow.com/a/43866992/14073862
```
Comment (mhofman) [inline packages/access-token/src/access-token.js]: Nit: I don't think source attribution belongs in the JSDoc, maybe keep the `//` leading comment for it?

File: packages/access-token/src/json-store.js [hunk 8]
```diff
@@ -218,7 +244,9 @@ function makeJSONStore(dirPath, forceReset = false) {
    * (if you want to save them, call commit() first).
    */
   async function close() {
-    // Nothing to do here.
+    assertNotClosed();
```
Comment (mhofman) [inline packages/access-token/src/json-store.js:247]: I'm wondering, what are the semantics of `close` for our other DBs? Is it a no-op if already closed, or does it throw like here?

File: packages/access-token/tsconfig.json (hunk 4 above)
Comment (mhofman) [reply packages/access-token/tsconfig.json:5]: `path` is surprising, but `proper-lockfile` I could understand. See comment on that import, and revert this if it does the trick.

File: packages/access-token/src/json-store.js [hunk 9]
```diff
@@ -2,6 +2,7 @@
 import fs from 'fs';
 import path from 'path';
 import process from 'process';
+import lockfile from 'proper-lockfile';
```
Comment (mhofman) [inline packages/access-token/src/json-store.js:5]: I think this is where the synthetic TS error you're seeing comes from. The [types](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/proper-lockfile/index.d.ts) are stricter than the actual [implementation](https://github.com/moxystudio/node-proper-lockfile/blob/master/index.js).


```suggestion
import { lock } from 'proper-lockfile';
```

or 

```suggestion
import * as lockfile from 'proper-lockfile';
```

**Review by mhofman — APPROVED** [review]
Approved, given that the comments are addressed.

File: packages/access-token/package.json (hunk 6 above)
Comment (michaelfig) [reply packages/access-token/package.json]: Ah, yes, they should be.

File: packages/access-token/src/json-store.js (hunk 9 above)
Comment (michaelfig) [reply packages/access-token/src/json-store.js:5]: There are synthetic TS errors for every bare specifier `import` in this file, so I see no reason to try working around this particular instance.

File: packages/access-token/src/json-store.js (hunk 8 above)
Comment (michaelfig) [reply packages/access-token/src/json-store.js:247]: `swing-store` throws an error on multiple close.

File: packages/access-token/tsconfig.json (hunk 4 above)
Comment (michaelfig) [reply packages/access-token/tsconfig.json:5]: Doesn't do the trick, so I'm keeping it.

Comment (mhofman) [reply packages/access-token/tsconfig.json:5]: Oh I never realized that other packages have this option too! My bad

**Comment by michaelfig** [comment]
@mergifyio refresh
